examples/erc1155/lib.rs

```diff
@@ -0,0 +1,687 @@
+    ///
+    /// This should be called by a compliant ERC-1155 contract if the intended recipient is a smart
+    /// contract.
+    ///
+    /// If the smart contract implementing this interface accepts token transfers then it must
+    /// return `MAGIC_VALUE` from this function. To reject a transfer it must revert.
+    ///
+    /// Any callers must revert if they receive anything other than `MAGIC_VALUE` as a return
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_received(
+        &mut self,
+        operator: AccountId,
+        from: AccountId,
+        token_id: TokenId,
+        value: Balance,
+        data: Vec<u8>,
+    ) -> Vec<u8>;
+
+    /// Handle the receipt of multiple ERC-1155 tokens.
+    ///
+    /// This should be called by a compliant ERC-1155 contract if the intended recipient is a smart
+    /// contract.
+    ///
+    /// If the smart contract implementing this interface accepts token transfers then it must
+    /// return `BATCH_MAGIC_VALUE` from this function. To reject a transfer it must revert.
+    ///
+    /// Any callers must revert if they receive anything other than `BATCH_MAGIC_VALUE` as a return
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_batch_received(
+        &mut self,
+        operator: AccountId,
+        from: AccountId,
+        token_ids: Vec<TokenId>,
+        values: Vec<Balance>,
+        data: Vec<u8>,
+    );
+}
+
+#[ink::contract]
+mod erc1155 {
+    use super::*;
+
+    use ink_env::call::{
+        build_call,
+        utils::ReturnType,
+        ExecutionInput,
+        Selector,
+    };
+    use ink_prelude::collections::BTreeMap;
+
+    /// Indicate that a token transfer has occured.
+    ///
+    /// This must be emitted even if a zero value transfer occurs.
+    #[ink(event)]
+    pub struct TransferSingle {
+        #[ink(topic)]
+        operator: AccountId,
+        #[ink(topic)]
+        from: AccountId,
+        #[ink(topic)]
+        to: AccountId,
+        token_id: TokenId,
+        value: Balance,
+    }
+
+    /// Indicate that an approval event has happened.
+    #[ink(event)]
+    pub struct ApprovalForAll {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        operator: AccountId,
+        approved: bool,
+    }
+
+    /// Indicate that a token's URI has been updated.
+    #[ink(event)]
+    pub struct Uri {
+        value: ink_prelude::string::String,
+        #[ink(topic)]
+        token_id: TokenId,
+    }
+
+    /// An ERC-1155 contract.
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Contract {
+        /// Tracks the balances of accounts across the different tokens that they might be holding.
+        balances: BTreeMap<(AccountId, TokenId), Balance>,
+        /// Which accounts (called operators) have been approved to spend funds on behalf of an owner.
+        ///
+        /// Note that the mapping is Set<(Owner, Operator)>.
+        approvals: BTreeMap<(AccountId, AccountId), ()>,
```

Comment:
To make this more readable you could make this mapping strongly typed by introducing a
```rust
struct Approval {
    owner: AccountId,
    operator: AccountId,
}
```
type. Not important though.

Reply:
Same goes for the `balances` mapping even though it is more obvious there.